### PR TITLE
logging: Anomaly logging

### PR DESCRIPTION
**Pull request by jlucovsky**
This changeset adds anomaly logging to suricata for [issue 2282](https://redmine.openinfosecfoundation.org/issues/2282)

Anomaly logging is controlled via the anomaly section within eve-log.
There is a single option -- packethdr -- for including the packet header
in the anomaly.

The previous PR has been updated so all anomaly log records have event_type of anomaly and contain a subtype (packet, stream, unknown) and if unknown, the event code.

Suricata-verify PR: [#41](https://github.com/OISF/suricata-verify/pull/41)

Note: continuation of #3801

Example output:

## Packet header logging disabled
```
{"timestamp":"1969-12-31T16:00:09.000000-0800","flow_id":759863467114496,"pcap_cnt":57,"event_type":"anomaly","src_ip":"204.252.102.11","dest_ip":"206.65.98.18","proto":"ICMP","icmp_type":224,"icmp_code":37,"anomaly":{"type":"packet","event":"decoder.icmpv4.unknown_type"}}
{"timestamp":"1969-12-31T16:00:09.000000-0800","pcap_cnt":58,"event_type":"anomaly","src_ip":"208.21.2.184","src_port":0,"dest_ip":"10.1.1.99","dest_port":0,"proto":"UDP","anomaly":{"type":"packet","event":"decoder.udp.pkt_too_small"}}
{"timestamp":"1969-12-31T16:00:09.000000-0800","pcap_cnt":59,"event_type":"anomaly","anomaly":{"type":"packet","event":"decoder.ipv4.wrong_ip_version"}}
```
## Packet header logging enabled
```
{"timestamp":"1969-12-31T16:00:09.000000-0800","flow_id":1261831327383552,"pcap_cnt":57,"event_type":"anomaly","src_ip":"204.252.102.11","dest_ip":"206.65.98.18","proto":"ICMP","icmp_type":224,"icmp_code":37,"packethdr":"FF FF FF FF FF FF 01 01 01 01 01 01 08 00 45 00 00 54 8F B3 40 00 FD 01 8A 99 CC FC 66 0B CE 41 ","linktype":1,"anomaly":{"type":"packet","event":"decoder.icmpv4.unknown_type"}}
{"timestamp":"1969-12-31T16:00:09.000000-0800","pcap_cnt":58,"event_type":"anomaly","src_ip":"208.21.2.184","src_port":0,"dest_ip":"10.1.1.99","dest_port":0,"proto":"UDP","packethdr":"FF FF FF FF FF FF 01 01 01 01 01 01 08 00 45 00 00 3C C5 9E 40 00 FF 11 D7 E0 D0 15 02 B8 0A 01 ","linktype":1,"anomaly":{"type":"packet","event":"decoder.udp.pkt_too_small"}}
{"timestamp":"1969-12-31T16:00:09.000000-0800","pcap_cnt":59,"event_type":"anomaly","packethdr":"FF FF FF FF FF FF 01 01 01 01 01 01 08 00 B4 49 8D 5E 2A 31 C6 61 C6 2E AA 85 9D CB 86 8C 24 62 ","linktype":1,"anomaly":{"type":"packet","event":"decoder.ipv4.wrong_ip_version"}}
```

### CODE REVIEW

File: src/output-json-anomaly.c [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+typedef struct JsonAnomalyLogThread_ {
+    /** LogFileCtx has the pointer to the file and a mutex to allow multithreading */
+    LogFileCtx* file_ctx;
+    MemBuffer *json_buffer;
+    AnomalyJsonOutputCtx* json_output_ctx;
+} JsonAnomalyLogThread;
+
+static int AnomalyJson(ThreadVars *tv, JsonAnomalyLogThread *aft, const Packet *p)
+{
+    bool is_IP_pkt = PKT_IS_IPV4(p) || PKT_IS_IPV6(p);
```
Comment (victorjulien) [inline src/output-json-anomaly.c]: style still not 100% correct: we use lowercase_snake_case for vars.

**Comment by victorjulien** [comment]
@jasonish could you review the log format?

**Comment by jlucovsky** [comment]
Will adjust to snake case pending comments by @jasonish and results of your QA

**Comment by jasonish** [comment]
In other event types we have a "packet_info" object that contains the "linktype".  These anomaly events should probably put their "linktype" in the same place... For example:
```
{
  "timestamp": "2019-04-22T12:49:48.224208-0600",
  "flow_id": 1999258098821753,
  "in_iface": "wlp4s0",
  "event_type": "alert",
  "src_ip": "10.16.1.202",
  "src_port": 52442,
  "dest_ip": "10.16.1.4",
  "dest_port": 5000,
  "proto": "TCP",
  "alert": {
    "action": "allowed",
    "gid": 1,
    "signature_id": 2600001,
    "rev": 1,
    "signature": "SURICATA HTTP on unusual port",
    "category": "",
    "severity": 3
  },
  "http": {
    "hostname": "nas1.unx.ca",
    "http_port": 5000,
    "url": "/webapi/entry.cgi",
    "http_user_agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.86 Safari/537.36",
    "http_content_type": "application/json",
    "http_refer": "http://nas1.unx.ca:5000/",
    "http_method": "POST",
    "protocol": "HTTP/1.1",
    "status": 200,
    "length": 82
  },
  "app_proto": "http",
  "flow": {
    "pkts_toserver": 4,
    "pkts_toclient": 3,
    "bytes_toserver": 965,
    "bytes_toclient": 690,
    "start": "2019-04-22T12:49:47.354937-0600"
  },
  "payload": "",
  "payload_printable": "",
  "stream": 0,
  "packet": "ABEyF0nwrO1c4E7GCABFAAA04yZAAEAGQLAKEAHKChABBMzaE4iRf4liMvS8sIAQAfPTLgAAAQEICggtQcILPEqZ",
  "packet_info": {
    "linktype": 1
  }
}
```
Note the "packet_info" object at the bottom.

**Comment by jasonish** [comment]
Regarding the packethdr option...

There is a trailing whitespace, can that be removed?

Its obviously a space separated hex string.  I can't recall anywhere else where we are logging a hex string.  @victorjulien Do you know of any?  I just want to make sure that we're consistent with formatting? For example, lower case vs upper case? Spaces? Prefixing with a 0x?  We have the flags as hex in netflow, but thats a little different than dumping binary data as hex.

If we don't have anything doing this already, then we should set the formatting rules here.

**Comment by victorjulien** [comment]
Yeah thinking about it, it would make a lot more sense to treat the packethdr the same as the 'packet' field: base64 encoded data.

**Comment by jasonish** [comment]
> Yeah thinking about it, it would make a lot more sense to treat the packethdr the same as the 'packet' field: base64 encoded data.

I'd be happy with base64.  Was just thinking if hex did make more sense for this (I'm not sure if it does?), then we should settle on a common format for hex strings.

**Comment by pevma** [comment]
I find hex easier to `read` when needed.

File: src/output-json-anomaly.c (hunk 1 above)
Comment (jlucovsky) [reply src/output-json-anomaly.c]: Will correct.

**Comment by jlucovsky** [comment]
My preference would be for hex (for humans) but I'll modify to use base64 encoding.

**Comment by jlucovsky** [comment]
Continued in #3819

**Pull request closed without merging**